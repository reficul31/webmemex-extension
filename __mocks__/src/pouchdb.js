import PouchDB from 'pouchdb-node'
import PouchDBMemory from 'pouchdb-adapter-memory'

PouchDB.plugin(PouchDBMemory)

const pouchdbOptions = {
    name: 'testdb',
    auto_compaction: true,
    adapter: 'memory',
}

const db = PouchDB(pouchdbOptions)
export default db
