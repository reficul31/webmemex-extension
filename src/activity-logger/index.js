// Stuff that is to be accessible from other modules (folders)

import docuri from 'docuri'
import randomString from 'src/util/random-string'

export const visitKeyPrefix = 'visit/'

/**
* Creates an _id string given the variables, or vice versa parses such strings
* We simply use the creation time for the id, for easy chronological sorting.
* We add a random string we call a 'nonce' to prevent accidental collisions.
*
* @param {Number} timestamp - timestamp of the document
* @param {Number} nonce - random 10 digit number
*
* @returns {String} - string id of the document
*/
export const convertVisitDocId = docuri.route(`${visitKeyPrefix}:timestamp/:nonce`)

/**
* Extract the type, timestamp and nonce from an ID of a document
*
* @param {String} - string id of the document
*/
const convertAnyTimestampedDocId = docuri.route(':type/:timestamp/:nonce')
export const getTimestamp = doc =>
    Number.parseInt(convertAnyTimestampedDocId(doc._id).timestamp)

/**
* Generate a visit document ID from the timestamp and a random string
*
* @param {Number} timestamp - timestamp of the visit document
* @param {Number} nonce - random 10 digit number
*
* @returns {string} - document id of the visit document
*/
export function generateVisitDocId({timestamp, nonce} = {}) {
    const date = timestamp ? new Date(timestamp) : new Date()
    return convertVisitDocId({
        timestamp: date.getTime(),
        nonce: nonce || randomString(),
    })
}

/**
* Tell whether a page can be stored
*
* @param {string} url - url of the visit document
*
* @returns {boolean} - whether or not the page is loggable
*/
export function isLoggable({url}) {
    // Only http(s) pages. Ignoring data URLs, newtab, ...
    const loggableUrlPattern = /^https?:\/\//
    return loggableUrlPattern.test(url)
}

/**
* Tell whether a page should be stored
*
* @param {string} url - url of the visit document
*
* @returns {boolean} - whether or not the page should be stored
*/
export function shouldBeLogged({url}) {
    // Currently, we log everything that we think we can log.
    // TODO Add ignore-/blacklist lookup here (issue #22, #94)
    return isLoggable({url})
}
