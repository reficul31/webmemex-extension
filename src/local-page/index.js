import get from 'lodash/fp/get'

/**
* Extract the local page URL from the page instance.
*
* @param {Page} page - stored page in the database
*
* @returns {string} - URL used to refer to the local page
*/
export function hrefForLocalPage({page}) {
    // Check if it has a stored page attached at all.
    if (!get(['_attachments', 'frozen-page.html'])(page)) {
        return undefined
    }
    const pageId = page._id
    const url = `/local-page.html?page=${encodeURIComponent(pageId)}`
    const hash = (page.url && page.url.split('#')[1])
    const href = (hash !== undefined) ? url + '#' + hash : url
    return href
}
