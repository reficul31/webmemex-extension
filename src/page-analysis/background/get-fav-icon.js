import responseToDataUrl from 'response-to-data-url'

/**
* Get a tab's fav-icon (website logo) as a data URL
*
* @param {Number} tabId - tabId of the current visit
*
* @returns {string} - favicon of the page as a data URL
*/
async function getFavIcon({tabId}) {
    const tab = await browser.tabs.get(tabId)

    if (tab.favIconUrl === undefined) {
        return undefined
    }

    const response = await fetch(tab.favIconUrl)
    const dataUrl = await responseToDataUrl(response)
    return dataUrl
}

export default getFavIcon
