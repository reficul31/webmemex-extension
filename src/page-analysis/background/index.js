import merge from 'lodash/fp/merge'
import { dataURLToBlob } from 'blob-util'

import { whenPageDOMLoaded } from 'src/util/tab-events'
import { remoteFunction } from 'src/util/webextensionRPC'
import whenAllSettled from 'when-all-settled'
import db from 'src/pouchdb'
import updateDoc, { setAttachment } from 'src/util/pouchdb-update-doc'

import { revisePageFields } from '..'
import getFavIcon from './get-fav-icon'
import makeScreenshot from './make-screenshot'

/**
* Extract page content and freeze-dry the page
*
* @param {string} pageId - page ID of the current page
* @param {Number} tabId - tabId of the current visit
*/
async function performPageAnalysis({pageId, tabId}) {
    // Run these functions in the content script in the tab.
    const extractPageContent = remoteFunction('extractPageContent', {tabId})
    const freezeDry = remoteFunction('freezeDry', {tabId})

    // Get and store the fav-icon
    const storeFavIcon = getFavIcon({tabId}).then(async dataUrl => {
        if (dataUrl === undefined) return
        const blob = await dataURLToBlob(dataUrl)
        await setAttachment(db, pageId, 'favIcon', blob)
    })

    // Capture a screenshot.
    const storeScreenshot = makeScreenshot({tabId}).then(async dataUrl => {
        if (dataUrl === undefined) return
        const blob = await dataURLToBlob(dataUrl)
        await setAttachment(db, pageId, 'screenshot', blob)
    })

    // Extract the text and metadata
    const storePageContent = extractPageContent().then(async content => {
        // Add the info to the doc's (possibly already existing) doc.content.
        await updateDoc(db, pageId, doc => merge({content})(doc))
    })

    // Freeze-dry and store the whole page
    async function storePageFreezeDried() {
        const htmlString = await freezeDry()
        const blob = new Blob([htmlString], {type: 'text/html;charset=UTF-8'})
        await setAttachment(db, pageId, 'frozen-page.html', blob)
    }

    // When every task has either completed or failed, update the search index.
    await whenAllSettled([
        storeFavIcon,
        storeScreenshot,
        storePageContent,
        storePageFreezeDried(),
    ])
}

/**
* Waits for the DOM to load and then stores and freeze-dries the page
*
* @param {Page} page - page instance of the current page
* @param {Number} tabId - tabId of the current visit
*/
export default async function analysePage({page, tabId}) {
    // Wait until its DOM has loaded, in case we got invoked before that.
    await whenPageDOMLoaded({tabId}) // TODO: catch e.g. tab close.
    await performPageAnalysis({pageId: page._id, tabId})
    // Get and return the page.
    page = revisePageFields(await db.get(page._id))
    return {page}
}
