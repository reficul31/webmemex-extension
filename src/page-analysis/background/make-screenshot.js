import delay from 'src/util/delay'
import { whenTabActive } from 'src/util/tab-events'

/**
* Take a screenshot of the tabId, if it is active.
* Returns a promise of the screenshot (a png image in a data URL).
* The promise rejects if the tab is not currently active!
*
* @param {Number} tabId - tabId of the current visit
*
* @returns {Image} - snapshot of the current page
*/
async function snapNow({tabId}) {
    const tab = await browser.tabs.get(tabId)
    const image = await browser.tabs.captureVisibleTab(
        tab.windowId,
        {format: 'png'}
    )
    return image
}

/**
* Return the promise of an image (as data URL) of the visible area of the tab,
* but only as soon as it is active (due to a limitation of the browser API)
*
* @param {Number} tabId - tabId of the current visit
*
* @returns {Image} - snapshot of the current page
*/
export default async function makeScreenshotOfTabAsap({tabId}) {
    await whenTabActive({tabId})
    // Some delay appears required to not fail. Perhaps the browser needs
    // to complete some rendering before the screen is captured?
    await delay(300)
    const image = await snapNow({tabId})
    return image
}
