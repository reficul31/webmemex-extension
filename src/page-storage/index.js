import docuri from 'docuri'
import randomString from 'src/util/random-string'

export const pageKeyPrefix = 'page/'

export const convertPageDocId = docuri.route(`${pageKeyPrefix}:timestamp/:nonce`)

/**
* Generate a page doc id
*
* @param {Number} timestamp - timestamp of the page
* @param {string} nonce - random string of 10 digit number
*
* @returns {string} - doc id of the page as a string
*/
export function generatePageDocId({timestamp, nonce} = {}) {
    const date = timestamp ? new Date(timestamp) : new Date()
    return convertPageDocId({
        timestamp: date.getTime(),
        nonce: nonce || randomString(),
    })
}
