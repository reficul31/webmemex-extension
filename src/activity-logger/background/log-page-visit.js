import db from 'src/pouchdb'
import { reidentifyOrStorePage } from 'src/page-storage/store-page'
import { makeRemotelyCallable } from 'src/util/webextensionRPC'
import { whenPageLoadComplete, whenTabActive } from 'src/util/tab-events'
import delay from 'src/util/delay'

import { generateVisitDocId, isLoggable, shouldBeLogged } from '..'

/**
* Store the visit in PouchDB and return the visit instance as an object
*
* @param {Number} timestamp - timestamp of the visit
* @param {string} url - URL of the visit document
* @param {Page} page - page studb of the visit document
*
* @returns {Visit} - instance of the visit
*/
async function storeVisit({timestamp, url, page}) {
    const visitId = generateVisitDocId({timestamp})
    const visit = {
        _id: visitId,
        visitStart: timestamp,
        url,
        page: {_id: page._id}, // store only a reference to the page
    }
    await db.put(visit)
    return {visit}
}

/**
* Log page visit and create a visit instance in the database
*
* @param {Number} tabId - tabId of the current visit
* @param {string} url - URL of the visit document
*
* @returns {Visit} - instance of the visit
* @returns {Page} - instance of the Page
*/
export async function logPageVisit({
    tabId,
    url,
}) {
    // The time to put in documents.
    const timestamp = Date.now()

    // TODO first try to extend an existing visit instead of logging a new one.

    // First create an identifier for the page being visited.
    const {page, finalPagePromise} = await reidentifyOrStorePage({tabId, url})
    // Create a visit pointing to this page (analysing/storing it may still be in progress)
    const visit = await storeVisit({page, url, timestamp})

    // Wait until all page analyis/deduping is done before returning.
    const {page: finalPage} = await finalPagePromise

    // TODO possibly deduplicate the visit if the page was deduped too.

    return {visit, page: finalPage}
}

/**
* Checks if logging is enabled and if enabled, logs the page
*
* @param {Number} tabId - tabId of the current visit
* @param {string} url - URL of the visit document
*
* @returns {Visit} - instance of the visit
* @returns {Page} - instance of the Page
*/
export async function maybeLogPageVisit({
    tabId,
    url,
}) {
    // Check if we would want to log this page.
    if (!shouldBeLogged({url})) {
        return
    }

    // Check if logging is enabled by the user.
    const { loggingEnabled } = await browser.storage.local.get('loggingEnabled')
    if (!loggingEnabled) {
        return
    }

    // Wait for it to be fully ready.
    await whenPageLoadComplete({tabId})
    // Wait a bit to first let scripts run. TODO Do this in a smarter way.
    await delay(1000)
    // Wait until the tab is activated (not technically needed)
    await whenTabActive({tabId})

    return await logPageVisit({
        tabId,
        url,
    })
}

/**
* Log the visit/page in the currently active tab
*
* @returns {Visit} - instance of the visit
* @returns {Page} - instance of the Page
*/
export async function logActivePageVisit() {
    const tabs = await browser.tabs.query({active: true, currentWindow: true})
    const {url, id: tabId} = tabs[0]

    if (!isLoggable({url})) {
        throw new Error('This page cannot be stored.')
    }

    return await logPageVisit({
        tabId,
        url,
    })
}

// Expose to be callable from browser button popup
makeRemotelyCallable({logActivePageVisit})
