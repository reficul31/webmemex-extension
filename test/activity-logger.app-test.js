/* eslint-env jest */
/* eslint import/namespace: "off" */

import { maybeLogPageVisit } from 'src/activity-logger/background/log-page-visit'
import { dataURLToBlob } from 'blob-util'
import * as webextensionRPC from 'src/util/webextensionRPC'
import db from 'src/pouchdb'
import * as activityLogger from 'src/activity-logger'

const imageDataUri = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAAAAAA6fptVAAAACklEQVR4nGNiAAAABgADNjd8qAAAAABJRU5ErkJggg=='

jest.mock('src/util/webextensionRPC')
jest.mock('src/util/tab-events')

browser.tabs = {
    get: jest.fn().mockReturnValue({
        windowId: 1,
        favIconUrl: 'https://example.com/icon.ico',
    }),
    captureVisibleTab: jest.fn().mockReturnValue(imageDataUri),
}
browser.storage = {
    local: {
        get: jest.fn().mockReturnValue({loggingEnabled: true,
        }),
    },
}

function removeRevisionKeys(page) {
    delete page._id
    delete page._rev
}

beforeEach(() => {
    const freezeDry = jest.fn().mockReturnValue(
        new Promise((resolve, reject) => {
            resolve('<html></html>')
        })
    )
    const extractPageContent = jest.fn().mockReturnValue(
        new Promise((resolve, reject) => {
            resolve('page content')
        })
    )
    webextensionRPC.remoteFunction = jest.fn().mockReturnValueOnce(extractPageContent).mockReturnValueOnce(freezeDry)
    activityLogger.generateVisitDocId = jest.fn().mockReturnValue('visit/1234567890123/1234567890')
})

afterAll(async () => {
    await db.destroy()
})

describe('App Test', () => {
    const tabId = 1
    const url = 'https://example.com/page'
    let favIcon
    let visit

    beforeAll(async () => {
        favIcon = await dataURLToBlob(imageDataUri)
    })

    test('should create a visit in the database', async () => {
        fetch.mockResponse(favIcon)
        await maybeLogPageVisit({tabId, url})
        visit = await db.get('visit/1234567890123/1234567890')
        expect(visit._id).toBe('visit/1234567890123/1234567890')
        expect(visit.url).toBe(url)
    })

    test('should create a page in the database for the visit', async () => {
        const page = await db.get(visit.page._id)
        removeRevisionKeys(page)
        expect(page).toMatchSnapshot()
    })
})
