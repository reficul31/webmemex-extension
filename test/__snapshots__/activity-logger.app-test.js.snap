// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Test should create a page in the database for the visit 1`] = `
Object {
  "_attachments": Object {
    "favIcon": Object {
      "content_type": "image/png",
      "digest": "md5-7cVkAmmX4suBnAFSJ4A2Wg==",
      "length": 67,
      "revpos": 4,
      "stub": true,
    },
    "frozen-page.html": Object {
      "content_type": "text/html;charset=utf-8",
      "digest": "md5-yDMBQlsq0dSWRzpf89nsyg==",
      "length": 13,
      "revpos": 3,
      "stub": true,
    },
    "screenshot": Object {
      "content_type": "image/png",
      "digest": "md5-7cVkAmmX4suBnAFSJ4A2Wg==",
      "length": 67,
      "revpos": 5,
      "stub": true,
    },
  },
  "content": "page content",
  "url": "https://example.com/page",
}
`;
